File: app/models/whatsapp.py
# app/models/whatsapp.py
from __future__ import annotations
from typing import List, Dict, Any, Optional, Union
from pydantic import BaseModel, Field
from datetime import datetime, timezone

# --- Menus ---

class WhatsAppMenu(BaseModel):
    tenant: str
    menu_id: str
    name: str
    status: str = "draft"  # draft | published
    version: int = 1
    tree: Dict[str, Any] = Field(default_factory=dict)
    locales: Dict[str, Any] = Field(default_factory=dict)
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_by: Optional[str] = None
    published_at: Optional[datetime] = None
    published_by: Optional[str] = None

class WhatsAppMenuUpsert(BaseModel):
    menu_id: str
    name: str
    tree: Dict[str, Any]
    locales: Optional[Dict[str, Any]] = None

class WhatsAppMenuListResponse(BaseModel):
    items: List[WhatsAppMenu]
    total: int

# --- Triggers ---

class TriggerMatch(BaseModel):
    type: str  # exact | prefix | contains | regex
    value: str
    locale: Optional[str] = None

class TriggerAction(BaseModel):
    kind: str  # render_submenu | jump_node | static_text | invoke_action
    menu_id: Optional[str] = None
    node_id: Optional[str] = None
    text: Optional[Union[str, Dict[str, str]]] = None

class WhatsAppTrigger(BaseModel):
    tenant: str
    trigger_id: str
    match: TriggerMatch
    action: TriggerAction
    enabled: bool = True
    priority: int = 0
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_by: Optional[str] = None

class WhatsAppTriggerCreate(BaseModel):
    trigger_id: str
    match: TriggerMatch
    action: TriggerAction
    enabled: bool = True
    priority: int = 0

class WhatsAppTriggerListResponse(BaseModel):
    items: List[WhatsAppTrigger]
    total: int

# --- Actions ---

class WhatsAppActionMeta(BaseModel):
    id: str
    label: str
    module: str
    requires_caps: List[str] = Field(default_factory=list)

class WhatsAppActionListResponse(BaseModel):
    items: List[WhatsAppActionMeta]
    total: int

# --- Bot simulation ---

class BotNextStepRequest(BaseModel):
    phone: str
    input: str
    menu_id: Optional[str] = None
    node: Optional[str] = None
    locale: Optional[str] = "en"

class BotNextStepResponse(BaseModel):
    reply: str
    node: Optional[str] = None
    session: Optional[Dict[str, Any]] = None

File: admin_ui/src/api/whatsapp.ts
import { api } from './axios'

export type WhatsAppMenu = {
  tenant: string
  menu_id: string
  name: string
  status: 'draft' | 'published'
  version?: number
  tree?: any
  locales?: Record<string, any>
  updated_at?: string
  updated_by?: string
  published_at?: string
  published_by?: string
}

export async function listMenus(tenant: string): Promise<{ items: WhatsAppMenu[]; total: number }>{
  const res = await api.get(`/tenants/${encodeURIComponent(tenant)}/whatsapp/menus`)
  return res.data
}

export async function getMenu(
  tenant: string,
  menuId: string,
  status?: 'draft' | 'published',
  version?: number,
): Promise<WhatsAppMenu>{
  const url = `/tenants/${encodeURIComponent(tenant)}/whatsapp/menus/${encodeURIComponent(menuId)}`
  const params: any = {}
  if (status) params.status = status
  if (typeof version === 'number') params.version = version
  const res = await api.get(url, { params: Object.keys(params).length ? params : undefined })
  return res.data
}

export async function upsertMenu(tenant: string, payload: { menu_id: string; name: string; tree: any; locales?: any }): Promise<WhatsAppMenu>{
  const res = await api.post(`/tenants/${encodeURIComponent(tenant)}/whatsapp/menus`, payload)
  return res.data
}

export async function publishMenu(tenant: string, menuId: string): Promise<WhatsAppMenu>{
  const res = await api.post(`/tenants/${encodeURIComponent(tenant)}/whatsapp/menus/${encodeURIComponent(menuId)}/publish`)
  return res.data
}

export async function deleteMenu(tenant: string, menuId: string): Promise<{ ok: boolean }>{
  const res = await api.delete(`/tenants/${encodeURIComponent(tenant)}/whatsapp/menus/${encodeURIComponent(menuId)}`)
  return res.data
}

// -------- Triggers API --------

export type TriggerMatch = {
  type: 'exact' | 'prefix' | 'contains' | 'regex'
  value: string
  locale?: string
}

export type TriggerAction =
  | { kind: 'render_submenu'; menu_id: string; node_id?: string }
  | { kind: 'jump_node'; menu_id: string; node_id: string }
  | { kind: 'static_text'; text: string | Record<string, string> }
  | { kind: 'invoke_action'; action_id: string }

export type WhatsAppTrigger = {
  tenant: string
  trigger_id: string
  match: TriggerMatch
  action: TriggerAction
  enabled: boolean
  priority: number
  updated_at?: string
  updated_by?: string
}

export async function listTriggers(tenant: string): Promise<{ items: WhatsAppTrigger[]; total: number }>{
  const res = await api.get(`/tenants/${encodeURIComponent(tenant)}/whatsapp/triggers`)
  return res.data
}

export async function getTrigger(tenant: string, trigger_id: string): Promise<WhatsAppTrigger>{
  const res = await api.get(`/tenants/${encodeURIComponent(tenant)}/whatsapp/triggers`)
  const items = (res.data?.items || []) as WhatsAppTrigger[]
  const found = items.find(t => t.trigger_id === trigger_id)
  if (!found) throw new Error('Trigger not found')
  return found
}

export async function createTrigger(tenant: string, trigger: Omit<WhatsAppTrigger, 'tenant' | 'updated_at' | 'updated_by'>): Promise<WhatsAppTrigger>{
  const res = await api.post(`/tenants/${encodeURIComponent(tenant)}/whatsapp/triggers`, trigger)
  return res.data
}

export async function updateTrigger(tenant: string, trigger_id: string, patch: Partial<WhatsAppTrigger>): Promise<WhatsAppTrigger>{
  const res = await api.patch(`/tenants/${encodeURIComponent(tenant)}/whatsapp/triggers/${encodeURIComponent(trigger_id)}`, patch)
  return res.data
}

export async function deleteTrigger(tenant: string, trigger_id: string): Promise<{ ok: boolean }>{
  const res = await api.delete(`/tenants/${encodeURIComponent(tenant)}/whatsapp/triggers/${encodeURIComponent(trigger_id)}`)
  return res.data
}

// -------- Actions Registry --------

export type WhatsAppActionMeta = {
  id: string
  label: string
  module: string
  requires_caps: string[]
}

export async function listAvailableActions(tenant?: string): Promise<WhatsAppActionMeta[]>{
  const res = await api.get(`/whatsapp/actions`, { params: tenant ? { tenant } : undefined })
  return res.data?.items || []
}

// -------- Tenant custom actions (reusable) --------

export type TenantCustomAction = {
  tenant?: string
  action_id: string
  name: string
  action_type: 'static_text' | 'predefined' | 'workflow'
  text?: string
  system_action_id?: string
  workflow_id?: string
  params?: Record<string, unknown>
  enabled?: boolean
  updated_at?: string
  updated_by?: string
}

export async function listCustomActions(tenant: string): Promise<{ items: TenantCustomAction[]; total: number }>{
  const res = await api.get(`/tenants/${encodeURIComponent(tenant)}/whatsapp/custom-actions`)
  return res.data
}

export async function upsertCustomAction(tenant: string, payload: TenantCustomAction): Promise<TenantCustomAction>{
  const res = await api.post(`/tenants/${encodeURIComponent(tenant)}/whatsapp/custom-actions`, payload)
  return res.data
}

export async function deleteCustomAction(tenant: string, actionId: string): Promise<{ ok: boolean }>{
  const res = await api.delete(`/tenants/${encodeURIComponent(tenant)}/whatsapp/custom-actions/${encodeURIComponent(actionId)}`)
  return res.data
}

export const PLACEHOLDER_HINTS = [
  '{{business_name}}', '{{name}}', '{{phone}}', '{{service}}',
  '{{professional}}', '{{date}}', '{{time}}', '{{appointment_id}}',
]

// Simple helper to test a phrase via dummy Twilio webhook
export async function testTriggerWebhook(toNumber: string, body: string): Promise<string>{
  const res = await api.post(`/integrations/twilio/whatsapp/webhook`, { From: '+911111111111', To: toNumber, Body: body }, { headers: { 'Content-Type': 'application/json' }})
  // Twilio webhook returns XML; axios by default parses text if not JSON
  return typeof res.data === 'string' ? res.data : JSON.stringify(res.data)
}

export async function botNextStep(tenant: string, payload: { phone: string; input: string; menu_id?: string; node?: string; locale?: string; reset_session?: boolean }): Promise<{ reply: string; node: string }>{
  const res = await api.post(`/bot/whatsapp/next`, payload, {
    headers: { 'X-Tenant': tenant }
  })
  return res.data
}
